File: leet_0659_make_consecutive_subseq___2maps.go
func isPossible(nums []int) bool {
    freq := make(map [int] int)
    subs := make(map [int] int)
    for _, n := range nums {
        freq[n]++
    }
    for _, n := range nums {
        if freq[n] == 0 {
            continue
        }
        freq[n]--
        if subs[n - 1] > 0 {
            subs[n - 1]--
            subs[n]++
        } else if freq[n + 1] > 0 && freq[n + 2] > 0 {
            freq[n + 1]--
            freq[n + 2]--
            subs[n + 2]++
        } else {
            return false
        }
    }
    return true
}

File: leet_0659_make_consecutive_subseq___2maps.cpp
class Solution {
public:
    bool isPossible(vector<int>& nums)
    {
        unordered_map<int, int>     freq;
        unordered_map<int, int>     subs;

        for (int n : nums)
            freq[n]++;
        for (int n : nums)
        {
            if (!freq[n])
                continue ;
            freq[n]--;
            if (subs[n - 1] > 0)
            {
                subs[n - 1]--;
                subs[n]++;
            }
            else if (freq[n + 1] > 0 && freq[n + 2] > 0)
            {
                freq[n + 1]--;
                freq[n + 2]--;
                subs[n + 2]++;
            }
            else
                return false;
        }
        return true;
    }
};
